### ISSUE TITLE
"SQL connection has changed" warning only solved with high wait_timeout

### ISSUE THREAD

**Issue by InvisibleFunction** [awslabs/aws-mysql-jdbc#372]
### Describe the bug

I had previously been using the standard MySQL Connector/J driver without issue and I recently switched to the AWS JDBC Driver. As soon as I did I started seeing these warnings in my logs:

```
WARN  pool.PoolBase - dbpool - Failed to validate connection software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ConnectionImpl@32e483dd (The active SQL connection has changed due to a connection failure. Please re-configure session state if required.)
```

My first thought was that these were caused by the db server's `wait_timeout` value of 10 seconds. I appended `sessionVariables=wait_timeout=28000` to my jdbc url and the warnings stopping being logged.

Here is the config for my Hikari datapool:

maximumPoolSize=25
minimumIdle=3
maxLifetime=115000
idleTimeout=20000
keepaliveTime=30000

Next, I set the `wait_timeout` on the connection url to 120. I would expect this configuration to work without issue, but instead I'm again getting the warning that I was getting before. Why does the only way that I can get that warning to stop seem to be to set the session `wait_timeout` to a very high value?

### Expected Behavior

I would expect that a `wait_timeout` set above the `maxLifetime` setting would prevent these warning messages.

### Current Behavior

I get warning messages in my logs.

### Reproduction Steps

I don't have an easy way to reproduce this yet.

### The AWS JDBC Driver for MySQL version used

1.1.4

### JDK version used

openjdk version "11.0.18" 2023-01-17

### Operating System and version

Ubuntu 20.04

**Comment by aaronchung-bitquill** [comment]
Hi @InvisibleFunction 

Thanks for reaching out to us and raising this issue. We'll taking a look at this and keep you updated as we investigate this. 

Thank you for your patience!

**Comment by aaronchung-bitquill** [comment]
Hi @InvisibleFunction 

Would you kindly provide us with some driver logs? Also can you tell us what plugins you are using with the wrapper driver? 

Thank you

**Comment by DanielWeintraub** [comment]
@aaronchung-bitquill 

The only logs I see from the driver (by searching my logs for 'software.aws.rds') are the "Failed to validate connection" messages mentioned above. Are there more logs you would want to look at? Some debug logging I could turn on?

I'm not sure what you mean about plugins? Would that be something I would have to enable in the driver by changing the connection string? I am using the same connection string I was using with the default jdbc driver.

Thanks!

**Comment by InvisibleFunction** [comment]
@aaronchung-bitquill

The only logs I see from the driver (by searching my logs for 'software.aws.rds') are the "Failed to validate connection" messages mentioned above. Are there more logs you would want to look at? Some debug logging I could turn on?

I'm not sure what you mean about plugins? Would that be something I would have to enable in the driver by changing the connection string? I am using the same connection string I was using with the default jdbc driver.

Thanks!

**Comment by aaronchung-bitquill** [comment]
Hi @InvisibleFunction 

That's right. Was looking for driver debug logs. Here are some documentation on enabling them. [Link](https://github.com/awslabs/aws-mysql-jdbc#enable-logging)

For the plugins, the aws-mysql-jdbc driver allows for the use of various connection plugins to add extra functionality to the `Connection` object. There are connection plugins that we provide such as the Failover, and Enhanced Failover Monitoring plugins. By default those two are enabled unless the `connectionPluginFactories` parameter is set. 
If you'd like to learn more about them,  here is a [link](https://github.com/awslabs/aws-mysql-jdbc#connection-plugin-manager)

**Comment by InvisibleFunction** [comment]
I added `&Slf4JLogger&enablePacketDebug=true` to the connection string, but didn't see anything additionally logged.

I then added `&Slf4JLogger&gatherPerfMetrics=true&enablePacketDebug=true&useUsageAdvisor=true` and those two new things definitely added lines to the logs so I think the logging is working, but I'm not seeing anything useful logged.

Any advice on options that would be helpful?

Regarding plugins, I do not set the `connectionPluginFactories` in the connection string.

**Comment by InvisibleFunction** [comment]
My connection string currently looks like this for testing:

```
jdbc:mysql:aws://dbhost.example.com:3306/db_user?clusterInstanceHostPattern=?.xxx-west-2.rds.amazonaws.com&autoReconnect=true&useSSL=false&requireSSL=false&verifyServerCertificate=false&Slf4JLogger&gatherPerfMetrics=true&enablePacketDebug=true&useUsageAdvisor=true&sessionVariables=wait_timeout=120
```

I also bumped up the wait_timeout incrementally and was still getting warnings with it set to 14400. I doubled it again and that's when I was able to get them to stop.

**Comment by aaronchung-bitquill** [comment]
Hi @InvisibleFunction 

Can you if the connection string is correct? 

The clusterInstanceHostPattern url parameter value is 
`clusterInstanceHostPattern=?.xxx-west-2.rds.amazonaws.com`

But i think it should be...
`clusterInstanceHostPattern=?.xxx.west-2.rds.amazonaws.com`

Thanks!

**Comment by InvisibleFunction** [comment]
You are correct, I made a mistake while posting and have corrected it in the above post. It does match what you say it should.

**Comment by aaronchung-bitquill** [comment]
Hi @InvisibleFunction 

Could you provide what version of Hikari you are using?

Additionally, could you take a look at our documentation on connection pooling and implement something similar to the custom exception `HikariCPSQLException` in the following sample code? [link](https://github.com/awslabs/aws-mysql-jdbc#connection-pooling)
Then set your `HikariConfig`  to use it like  `config.setExceptionOverrideClassName(HikariCPSQLException.class.getName());`

While this doesnt solve the underlying issue, this should alleviate the symptoms in the connection pool behaviour. 

Thank you!

**Comment by InvisibleFunction** [comment]
> Could you provide what version of Hikari you are using?

Looks like we’re on 3.4.5

**Comment by aaronchung-bitquill** [comment]
Hi @InvisibleFunction 

I think you may have the logger parameter set incorrectly. I believe it needs to be set like `logger=Slf4JLogger`
If that does not work, can you try `logger=StandardLogger`. It should print a ton of info to the console. 

Thank you

**Comment by aaronchung-bitquill** [comment]
Hi @InvisibleFunction 

Sorry for not getting back sooner. The recommended solution here is to implement  a `SQLExceptionOverride` and set the `exceptionOverrideClassName` configuration to it. There is a Hikari specific example here in our [documentation](https://github.com/awslabs/aws-mysql-jdbc#connection-pooling).

The reason for this is that when the `aws-mysql-jdbc` driver detects a connection failure and switches connection, it throws a SQLException with specific sqlState code. It does this to let the application know that a connection switch has happened, and it may need to set session settings information again. 

However, many connection pooling libraries have additional logic handling SQLExceptions and may not recognize the exceptions created by the `aws-mysql-jdbc` driver. In this case, the result is Hikari closing the connection objects prematurely. So, users of this driver are recommended to investigate their connection pool library and implement the required code to allow the connection pool to accept the exceptions raised by the driver. For Hikari, this is done by implementing the`SQLExceptionOverride` interface, and setting the `exceptionOverrideClassName` Hikari configuration to the implemented `SQLExceptionOverride` class name.

The below is an example of how to implement the `SQLExceptionOverride` interface.

```java
import com.zaxxer.hikari.SQLExceptionOverride;

import java.sql.SQLException;

public class HikariCPSQLException implements SQLExceptionOverride {
    public Override adjudicate(final SQLException sqlException) {
        String sqlState = sqlException.getSQLState();
        if (sqlState.equalsIgnoreCase("08S02") ||
            sqlState.equalsIgnoreCase("08007")) {
            return Override.DO_NOT_EVICT;
        } else {
            return Override.CONTINUE_EVICT;
        }
    }
}
``` 
_Note: `08S02` and `08007` are codes that identify SQLExceptions the `aws-mysql-jdbc` driver raises._ 

Setting the `exceptionOverrideClassName` configuration can be done in a number of ways. If you are using HikariConfig, it can be set using the `setExceptionOverrideClassName()` method like [here](setExceptionOverrideClassName). Or if you are using Spring Boot, it can be set in your `application.properties` file by setting the `spring.datasource.hikari.exception-override-class-name` field. 

Please let us know how this solution goes for you, or if you have any other questions/issues. 

Thank you

**Comment by InvisibleFunction** [comment]
Thank you, this is very helpful.

It sounds like you think that the errors and warnings I am getting regarding connection failures are in fact, connection failures (amazing). I'm not sure if my prior configuration was not experiencing these connection failures or if it was masking them.

Either way it sounds like we have some work to do to handle connection failures properly and I have some new avenues of investigation to see why we are experiencing so many connection failures.

**Comment by aaronchung-bitquill** [comment]
Hi @InvisibleFunction 

Checking back in again to see if you have any other issues/questions. Otherwise we will closing this ticket in the next few days. Note that if you have any issues/questions after we've closed this ticket, you can still re-open this ticket or create a new one. 

Thank you
